### PR TITLE
TosaToTensor: Support reshape on unsigned

### PR DESCRIPTION
**Pull request by mgehre-amd**
Update the conversion for tosa.reshape to properly use the type converter, so tosa.reshape on unsigned types work.

### CODE REVIEW

**Comment by cferry-AMD** [comment]
Is there an element-wise cast operator for tensors that would account for the unrealized casts? Besides, unless I'm mistaking, tensor reshaping [should not depend on the element type](https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorcollapse_shape-tensorcollapseshapeop), so I think I missed something here.

**Comment by mgehre-amd** [comment]
> Is there an element-wise cast operator for tensors that would account for the unrealized casts? Besides, unless I'm mistaking, tensor reshaping [should not depend on the element type](https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorcollapse_shape-tensorcollapseshapeop), so I think I missed something here.

When converting from TOSA to linalg, we need to convert unsigned to signless integers. This is done via the type converted, but all RewritePatterns need to correctly use it to determine the destination types.
The one for tosa.reshape didn't do that, so it would create linalg IR with unsigned types.

The casts are `unrealized casts` because they are expected to fold away, and not be realized into actual operations like tensor.cast.

File: mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp [hunk 1]
```diff
@@ -42,7 +43,10 @@ struct TosaToTensor : public impl::TosaToTensorBase<TosaToTensor> {
     target.addLegalDialect<arith::ArithDialect>();
     target.addLegalDialect<tensor::TensorDialect>();
 
-    mlir::tosa::populateTosaToTensorConversionPatterns(&patterns);
+    TypeConverter converter;
+    mlir::tosa::populateTosaToLinalgTypeConversion(converter);
```
Comment (cferry-AMD) [inline mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp:47]: Doesn't this make the TosaToTensor specifically targeted towards Linalg ops? I don't know of any other possible lowerings but still... I'd like to make sure.

**Review by cferry-AMD — APPROVED** [review]
This looks OK from a technical standpoint. I don't want my absence of knowledge of TOSA-to-{Linalg,Tensor} to block you for the rest.

File: mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp (hunk 1 above)
Comment (mgehre-amd) [reply mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp:47]: It uses the type converter from there, but it still only generate `tensor` ops. I don't know of a better place to put the type converter; maybe upstream will have an idea.

File: mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include "mlir/Pass/PassManager.h"
 #include "mlir/Transforms/DialectConversion.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include <mlir/Conversion/TosaToLinalg/TosaToLinalg.h>
```
Comment (TinaAMD) [inline mlir/lib/Conversion/TosaToTensor/TosaToTensorPass.cpp:23]:
```suggestion
#include "mlir/Conversion/TosaToLinalg/TosaToLinalg.h"
```
(and sort)

File: mlir/test/Conversion/TosaToTensor/tosa-to-tensor.mlir [hunk 3]
```diff
@@ -56,6 +56,19 @@ func.func @test_reshape_samerank(%arg0: tensor<3x2xf32>) -> tensor<2x3xf32> {
 
 // -----
 
+// CHECK-LABEL: @test_reshape_samerank_unsigned
+//  CHECK-SAME: (%[[ARG0:.*]]: tensor<3x2xui8>)
+func.func @test_reshape_samerank_unsigned(%arg0: tensor<3x2xui8>) -> tensor<2x3xui8> {
+  // CHECK-NEXT: %[[CAST1:.*]] = builtin.unrealized_conversion_cast %[[ARG0]] : tensor<3x2xui8> to tensor<3x2xi8>
+  // CHECK-NEXT: %[[RESHAPE1:.*]] = tensor.collapse_shape %[[CAST1]] {{\[}}[0, 1]] : tensor<3x2xi8> into tensor<6xi8>
+  // CHECK-NEXT: %[[RESHAPE2:.*]] = tensor.expand_shape %[[RESHAPE1]] {{\[}}[0, 1]] : tensor<6xi8> into tensor<2x3xi8>
+  // CHECK-NEXT: %[[CAST2:.*]] = builtin.unrealized_conversion_cast %[[RESHAPE2]] : tensor<2x3xi8> to tensor<2x3xui8
```
Comment (TinaAMD) [inline mlir/test/Conversion/TosaToTensor/tosa-to-tensor.mlir:65]:
```suggestion
  // CHECK-NEXT: %[[CAST2:.*]] = builtin.unrealized_conversion_cast %[[RESHAPE2]] : tensor<2x3xi8> to tensor<2x3xui8>
```

**Review by TinaAMD — APPROVED** [review]
LGTM

**Pull request merged**